# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [10]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [11]:
pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0             2018               1            367   
1             2018               1            367   
2             2018               1            367   
3             2018               1            367   
4             2018               1            367   
5             2018               1            367   
6             2018               1            367   
7             2018               1            367   
8             2018               1            367   
9             2018               1            367   
10            2018               1            367   
11            2018               1            367   
12            2018               1            367   
13            2018               1            367   
14            2018               1            367   
15            2018               1            367   
16            2018               1            367   
17            2018               1            367   
18            2018               1            367   
19            2018               1            367   
20            2018               1            367   
21            2018               1            367   
22            2018               1            367   
23            2018               1            367   
24            2018               1            367   
25            2018               1            367   
26            2018               1            367   
27            2018               1            367   
28            2018               1            367   
29            2018               1            367   
...            ...             ...            ...   
24064         2018               4            367   
24065         2018               4            367   
24066         2018               4            367   
24067         2018               4            367   
24068         2018               4            367   
24069         2018               4            367   
24070         2018               4            367   
24071         2018               4            367   
24072         2018               4            367   
24073         2018               4            367   
24074         2018               4            367   
24075         2018               4            367   
24076         2018               4            367   
24077         2018               4            367   
24078         2018               4            367   
24079         2018               4            367   
24080         2018               4            367   
24081         2018               4            367   
24082         2018               4            367   
24083         2018               4            367   
24084         2018               4            367   
24085         2018               4            367   
24086         2018               4            367   
24087         2018               4            367   
24088         2018               4            367   
24089         2018               4            367   
24090         2018               4            367   
24091         2018               4            367   
24092         2018               4            367   
24093         2018               4            367   

                  Business_Unit_Descr  GAAP_Fund_Type  \
0      Kansas State University                     12   
1      Kansas State University                     12   
2      Kansas State University                     12   
3      Kansas State University                     12   
4      Kansas State University                     12   
5      Kansas State University                     12   
6      Kansas State University                     12   
7      Kansas State University                     12   
8      Kansas State University                     12   
9      Kansas State University                     12   
10     Kansas State University                     12   
11     Kansas State University      

In [19]:
ksuexp = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [12]:
pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0            2018               1            367   
1            2018               1            367   
2            2018               1            367   
3            2018               1            367   
4            2018               1            367   
5            2018               1            367   
6            2018               1            367   
7            2018               1            367   
8            2018               1            367   
9            2018               1            367   
10           2018               1            367   
11           2018               1            367   
12           2018               1            367   
13           2018               1            367   
14           2018               1            367   
15           2018               1            367   
16           2018               1            367   
17           2018               1            367   
18           2018               1            367   
19           2018               1            367   
20           2018               1            367   
21           2018               1            367   
22           2018               1            367   
23           2018               1            367   
24           2018               1            367   
25           2018               1            367   
26           2018               1            367   
27           2018               1            367   
28           2018               1            367   
29           2018               1            367   
...           ...             ...            ...   
1880         2018               4            367   
1881         2018               4            367   
1882         2018               4            367   
1883         2018               4            367   
1884         2018               4            367   
1885         2018               4            367   
1886         2018               4            367   
1887         2018               4            367   
1888         2018               4            367   
1889         2018               4            367   
1890         2018               4            367   
1891         2018               4            367   
1892         2018               4            367   
1893         2018               4            367   
1894         2018               4            367   
1895         2018               4            367   
1896         2018               4            367   
1897         2018               4            367   
1898         2018               4            367   
1899         2018               4            367   
1900         2018               4            367   
1901         2018               4            367   
1902         2018               4            367   
1903         2018               4            367   
1904         2018               4            367   
1905         2018               4            367   
1906         2018               4            367   
1907         2018               4            367   
1908         2018               4            367   
1909         2018               4            367   

                 Business_Unit_Descr  GAAP_Fund_Type  \
0     Kansas State University                      1   
1     Kansas State University                     12   
2     Kansas State University                     12   
3     Kansas State University                     12   
4     Kansas State University                     12   
5     Kansas State University                     12   
6     Kansas State University                     12   
7     Kansas State University                     12   
8     Kansas State University                     12   
9     Kansas State University                     12   
10    Kansas State University                     12   
11    Kansas State University                     12   
12    Kansas State University                     12 

In [18]:
ksurev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [13]:
pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0             2018               1            682   
1             2018               1            682   
2             2018               1            682   
3             2018               1            682   
4             2018               1            682   
5             2018               1            682   
6             2018               1            682   
7             2018               1            682   
8             2018               1            682   
9             2018               1            682   
10            2018               1            682   
11            2018               1            682   
12            2018               1            682   
13            2018               1            682   
14            2018               1            682   
15            2018               1            682   
16            2018               1            682   
17            2018               1            682   
18            2018               1            682   
19            2018               1            682   
20            2018               1            682   
21            2018               1            682   
22            2018               1            682   
23            2018               1            682   
24            2018               1            682   
25            2018               1            682   
26            2018               1            682   
27            2018               1            682   
28            2018               1            682   
29            2018               1            682   
...            ...             ...            ...   
10523         2018               4            682   
10524         2018               4            682   
10525         2018               4            682   
10526         2018               4            682   
10527         2018               4            682   
10528         2018               4            682   
10529         2018               4            682   
10530         2018               4            682   
10531         2018               4            682   
10532         2018               4            682   
10533         2018               4            682   
10534         2018               4            682   
10535         2018               4            682   
10536         2018               4            682   
10537         2018               4            682   
10538         2018               4            682   
10539         2018               4            682   
10540         2018               4            682   
10541         2018               4            682   
10542         2018               4            682   
10543         2018               4            682   
10544         2018               4            682   
10545         2018               4            682   
10546         2018               4            682   
10547         2018               4            682   
10548         2018               4            682   
10549         2018               4            682   
10550         2018               4            682   
10551         2018               4            682   
10552         2018               4            682   

                  Business_Unit_Descr  GAAP_Fund_Type  \
0      University of Kansas                         1   
1      University of Kansas                         1   
2      University of Kansas                         1   
3      University of Kansas                         1   
4      University of Kansas                         1   
5      University of Kansas                         1   
6      University of Kansas                         1   
7      University of Kansas                         1   
8      University of Kansas                         1   
9      University of Kansas                         1   
10     University of Kansas                         1   
11     University of Kansas         

In [17]:
kuexp = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [15]:
pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0            2018               1            682   
1            2018               1            682   
2            2018               1            682   
3            2018               1            682   
4            2018               1            682   
5            2018               1            682   
6            2018               1            682   
7            2018               1            682   
8            2018               1            682   
9            2018               1            682   
10           2018               1            682   
11           2018               1            682   
12           2018               1            682   
13           2018               1            682   
14           2018               1            682   
15           2018               1            682   
16           2018               1            682   
17           2018               1            682   
18           2018               1            682   
19           2018               1            682   
20           2018               1            682   
21           2018               1            682   
22           2018               1            682   
23           2018               1            682   
24           2018               1            682   
25           2018               1            682   
26           2018               1            682   
27           2018               1            682   
28           2018               1            682   
29           2018               1            682   
...           ...             ...            ...   
1047         2018               4            682   
1048         2018               4            682   
1049         2018               4            682   
1050         2018               4            682   
1051         2018               4            682   
1052         2018               4            682   
1053         2018               4            682   
1054         2018               4            682   
1055         2018               4            682   
1056         2018               4            682   
1057         2018               4            682   
1058         2018               4            682   
1059         2018               4            682   
1060         2018               4            682   
1061         2018               4            682   
1062         2018               4            682   
1063         2018               4            682   
1064         2018               4            682   
1065         2018               4            682   
1066         2018               4            682   
1067         2018               4            682   
1068         2018               4            682   
1069         2018               4            682   
1070         2018               4            682   
1071         2018               4            682   
1072         2018               4            682   
1073         2018               4            682   
1074         2018               4            682   
1075         2018               4            682   
1076         2018               4            682   

                 Business_Unit_Descr  GAAP_Fund_Type  \
0     University of Kansas                        12   
1     University of Kansas                        12   
2     University of Kansas                        12   
3     University of Kansas                        12   
4     University of Kansas                        12   
5     University of Kansas                        12   
6     University of Kansas                        12   
7     University of Kansas                        12   
8     University of Kansas                        12   
9     University of Kansas                        12   
10    University of Kansas                        12   
11    University of Kansas                        12   
12    University of Kansas                        12 

In [16]:
kurev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [24]:
ksurev.sum()

Fiscal_Year                                                                  3854380
Fiscal_Quarter                                                                  4781
Business_Unit                                                                 700970
Business_Unit_Descr                Kansas State University       Kansas State Uni...
GAAP_Fund_Type                                                                 22825
GAAP_Fund_Type_Descr               General Fund                  Component Unit  ...
Fund_Code                                                                    7600145
Fund_Code_Descr                    STATE GENERAL FUND            NATIONAL BIO AGR...
Budget_Ref                                                                   6903830
Budget_Ref_Descr                   OPERATING EXP-INCLD OFF HOS                   ...
Program_Code                                                               108008432
Program_Code_Descr                 Fiscal Operating              

In [25]:
kurev.sum()

Fiscal_Year                                                                  2173386
Fiscal_Quarter                                                                  2782
Business_Unit                                                                 734514
Business_Unit_Descr                University of Kansas          University of Ka...
GAAP_Fund_Type                                                                 12924
GAAP_Fund_Type_Descr               Component Unit                Component Unit  ...
Fund_Code                                                                    4513144
Fund_Code_Descr                    GENERAL FF                    GENERAL FF      ...
Budget_Ref                                                                   4229699
Budget_Ref_Descr                   GENERAL FF                                    ...
Program_Code                                                                59202000
Program_Code_Descr                 Institutional Support         

In [26]:
3854380 - 2173386

1680994

<b>Answer:</b> KSU made $3854380, KU made $2,173,386. KSU made $1,680,994 more. 

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [27]:
ksuexp.sum()

Fiscal_Year                                                                 48621692
Fiscal_Quarter                                                                 61020
Business_Unit                                                                8842498
Business_Unit_Descr                Kansas State University       Kansas State Uni...
GAAP_Fund_Type                                                                252997
GAAP_Fund_Type_Descr               Component Unit                Component Unit  ...
Fund_Code                                                                   65114734
Fund_Code_Descr                    SMITH-LEVER SPEC PRG GRNT FDF SMITH-LEVER SPEC...
Budget_Ref                                                                  58450444
Budget_Ref_Descr                   SMITH-LEVER SPEC PRG GRNT FDF                 ...
Program_Code                                                              1165450828
Program_Code_Descr                 Cooperative Extension Services

In [42]:
kuexp.sum()

Fiscal_Year                                                                 21295954
Fiscal_Quarter                                                                 27075
Business_Unit                                                                7197146
Business_Unit_Descr                University of Kansas          University of Ka...
GAAP_Fund_Type                                                                115356
GAAP_Fund_Type_Descr               General Fund                  General Fund    ...
Fund_Code                                                                   29412141
Fund_Code_Descr                    STATE GENERAL FUND            STATE GENERAL FU...
Budget_Ref                                                                  26511953
Budget_Ref_Descr                   OPERATING EXP-INCLD OFF HOS                   ...
Program_Code                                                               549045000
Program_Code_Descr                 Instructional Services        

In [29]:
48621692 - 21295954

27325738

<b>Answer:</b> KSU spent $48,621,692 and KU spent $21,295,954. KSU spent $27,325,738 more.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [62]:
ksurev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [63]:
kurev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

<b>Answer:</b> They were from Gift Donation Federal Grants, Taxes, Revenue Tranfers, and more. 

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [31]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [69]:
ksurev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

<b>Answer:</b> Gift Donations and Federal Grants, same as KU. 

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [68]:
#ksurev["Primary_Account_Code_Descr"] = ksurev["Primary_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

<b>Answer:</b>

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [64]:
ksuexp.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    64350550.130
COMMODITIES                       42670908.840
CONTRACTUAL SERVICES             128541598.280
DEBT SERVICE                      32442445.100
EXPENSE TRANSFERS                 37130576.230
GRANTS CLAIMS SHARED REVENUE     106551248.640
NONEXPENSE ITEMS                 130529992.790
SALARIES AND WAGES               476367053.010
Name: Total, dtype: float64

In [65]:
kuexp.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    70676545.250
COMMODITIES                       24225857.120
CONTRACTUAL SERVICES             139630507.780
DEBT SERVICE                      25309742.030
EXPENSE TRANSFERS                 25815417.520
GRANTS CLAIMS SHARED REVENUE      71439705.360
NONEXPENSE ITEMS                 161145084.400
SALARIES AND WAGES               471996016.150
Name: Total, dtype: float64

<b>Answer:</b> Both universities spent money on Commodities, Contractual Services, Salaries and Wages, Expense Transfers, Debt Services, and more. 

### Q7: What costs two universities the most? 

In [70]:
ksuexp.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    64350550.130
COMMODITIES                       42670908.840
CONTRACTUAL SERVICES             128541598.280
DEBT SERVICE                      32442445.100
EXPENSE TRANSFERS                 37130576.230
GRANTS CLAIMS SHARED REVENUE     106551248.640
NONEXPENSE ITEMS                 130529992.790
SALARIES AND WAGES               476367053.010
Name: Total, dtype: float64

In [71]:
kuexp.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    70676545.250
COMMODITIES                       24225857.120
CONTRACTUAL SERVICES             139630507.780
DEBT SERVICE                      25309742.030
EXPENSE TRANSFERS                 25815417.520
GRANTS CLAIMS SHARED REVENUE      71439705.360
NONEXPENSE ITEMS                 161145084.400
SALARIES AND WAGES               471996016.150
Name: Total, dtype: float64

In [73]:
ksuexp.sort_values("Primary_Account_Code_Descr", ascending=False)

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
12047         2018               3            367   
13237         2018               3            367   
5669          2018               1            367   
13220         2018               3            367   
13221         2018               3            367   
13222         2018               3            367   
13223         2018               3            367   
13224         2018               3            367   
13225         2018               3            367   
13226         2018               3            367   
13227         2018               3            367   
13228         2018               3            367   
13229         2018               3            367   
13230         2018               3            367   
13231         2018               3            367   
13238         2018               3            367   
5671          2018               1            367   
13239         2018               3            367   
13240         2018               3            367   
5652          2018               1            367   
13297         2018               3            367   
5647          2018               1            367   
5646          2018               1            367   
5645          2018               1            367   
5644          2018               1            367   
5643          2018               1            367   
5642          2018               1            367   
5641          2018               1            367   
5640          2018               1            367   
5639          2018               1            367   
...            ...             ...            ...   
16561         2018               3            367   
9367          2018               2            367   
8243          2018               2            367   
17138         2018               3            367   
7397          2018               2            367   
22923         2018               4            367   
7396          2018               2            367   
11071         2018               2            367   
11070         2018               2            367   
4439          2018               1            367   
19032         2018               4            367   
7457          2018               2            367   
22984         2018               4            367   
22983         2018               4            367   
22982         2018               4            367   
22981         2018               4            367   
7456          2018               2            367   
4437          2018               1            367   
19031         2018               4            367   
19033         2018               4            367   
13293         2018               3            367   
19034         2018               4            367   
19035         2018               4            367   
19036         2018               4            367   
19037         2018               4            367   
4438          2018               1            367   
13290         2018               3            367   
13291         2018               3            367   
13292         2018               3            367   
5390          2018               1            367   

                  Business_Unit_Descr  GAAP_Fund_Type  \
12047  Kansas State University                      1   
13237  Kansas State University                     12   
5669   Kansas State University                     12   
13220  Kansas State University                     12   
13221  Kansas State University                     12   
13222  Kansas State University                     12   
13223  Kansas State University                     12   
13224  Kansas State University                     12   
13225  Kansas State University                     12   
13226  Kansas State University                     12   
13227  Kansas State University                     12   
13228  Kansas State University      

<b>Answer:</b> Salaries and Wages cost both universities the most.

### Q8: Calculate the percentage of this cost over each university's total expense.

In [75]:
471996016.150 / 1018584373 * 100

46.338430930355536

<b>Answer:</b> 46.34%

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

In [51]:
ksuexp.sum()

Fiscal_Year                                                                 48621692
Fiscal_Quarter                                                                 61020
Business_Unit                                                                8842498
Business_Unit_Descr                Kansas State University       Kansas State Uni...
GAAP_Fund_Type                                                                252997
GAAP_Fund_Type_Descr               Component Unit                Component Unit  ...
Fund_Code                                                                   65114734
Fund_Code_Descr                    SMITH-LEVER SPEC PRG GRNT FDF SMITH-LEVER SPEC...
Budget_Ref                                                                  58450444
Budget_Ref_Descr                   SMITH-LEVER SPEC PRG GRNT FDF                 ...
Program_Code                                                              1165450828
Program_Code_Descr                 Cooperative Extension Services

In [57]:
48621692 - 3854380 

44767312

In [55]:
kuexp.sum()

Fiscal_Year                                                                 21295954
Fiscal_Quarter                                                                 27075
Business_Unit                                                                7197146
Business_Unit_Descr                University of Kansas          University of Ka...
GAAP_Fund_Type                                                                115356
GAAP_Fund_Type_Descr               General Fund                  General Fund    ...
Fund_Code                                                                   29412141
Fund_Code_Descr                    STATE GENERAL FUND            STATE GENERAL FU...
Budget_Ref                                                                  26511953
Budget_Ref_Descr                   OPERATING EXP-INCLD OFF HOS                   ...
Program_Code                                                               549045000
Program_Code_Descr                 Instructional Services        

In [58]:
21295954 - 2173386 

19122568

<b>Answer:</b> KSU made $44,767,312 and KU made $19,122,568.


### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> Python

### END OF THE TEST